### ISSUE TITLE
Simplify branching, updating, approval and deployment workflow

### ISSUE THREAD

**Issue by hitchcott** [ethereumclassic/ethereumclassic.github.io#271]
*Sorry for the length of this; I wanted to be precise.*

~~TLDR: I'm going to look into using Github Actions for builds~~

UPDATE: Github Actions doesn't give us what we want, I suggest we use netlify.

Currently the branching system is a [little confusing and annoying](https://github.com/ethereumclassic/ethereumclassic.github.io/tree/staging#branches) for managing updates. 

Because we're using github pages and want to maintain a balance of security and openness (while using an automated build system that was designed without branch protection), we've ended up with a bit of an inelegant workflow for updating content.

## What We Have Now

First, a bit of backstory on why we have the existing approach.

The two branch types (compiled and source code) serve the purpose of maintaining source code and compiled deployment code separately, which is a [requirement](https://help.github.com/en/github/working-with-github-pages/about-github-pages#publishing-sources-for-github-pages-sites
) for main repository organisation github pages (i.e. etheruemclassic.github.io).

> The default publishing source for user and organization sites is the master branch. If the repository for your user or organization site has a master branch, your site will publish automatically from that branch. You cannot choose a different publishing source for user or organization sites.

In other words, for this repo, `master` cannot be used for source code, only for it's compiled production code. But no big deal; we can just use a `source` branch for source code and have TravisCI push the compiled code to to `master`.

However, we rightfully have enabled branch protection on `master` to prevent individuals from updating the deployed content. 3 different maintainers need to approve any PR to this branch. This is good for security, but has forced the existing awkward workaround:

Source code PRs are merged into `source`, Travis uses these commits to compile and push to `staging`, then `staging` is used to create PRs into `master` for approval. Whatever is on HEAD of `master` is shown on etheruemclassic.github.io (and thus etheruemclassic.org).

To review, we currently have:

- `master` (*compiled* code; is protected and requires 3 people to approve updates)
- `staging` (*compiled* code; Travis pushes here as master cannot be commit to directly) 
- `source` (*source* code; commits here are compiled by Travis and pushed to `staging`)
- Other branches and forks (*source* code, to be merged into `source`)

## The Problem

This system allows us to work within the github pages restrictions with a protected `master` branch, but there are some drawbacks:

* It is confusing
* Can't easily review the approved PRs into master (as compiled code is approved rather than source code)
* Critically, cannot approve multiple `source` PRs in parallel (as multiple PRs to `source` will update the `staging` branch and mess with unapproved PRs from `staging` into `master`)

In other words, the existing setup sucks. It is a hacky workaround.

## Suggested Solutions

So I tried to find some way to make it better. Ideally, I think we want to avoid having to deal with PRs involving compiled code and only manage source code PRs. Anything to do with compiled code should be handled by some CI; automatically building and deploying from approved source code.

To achieve this we either need to allow a CI to push directly to `master`, or use some other deployment solution than github pages. There isn't a clear winning solution as they all have one drawback or another, but I have explored a few options below.

### Possible Solution A: Keep Github Pages and have CI push directly to `master`

Solution A keeps the existing deployment setup including domain CNAME settings; the website is still hosted on github pages. The only way to achieve this without managing compiled code PRs is to have a CI tool push directly to `master`. 

For all of the following solutions, we'd enable branch protection on `source`. Anything approved on `source` gets automatically built and published. PR previews will still exist as they do now.

#### A1: Allow TravisCI to push directly to `master`

We can keep our existing CI tool (TravisCI) and instead of pushing to `staging` have it push directly to `master`. We no longer have to deal with compiled code PRs, at the expense of reduced security.

If it's possible to allow only TravisCI to publish to master, we could enable branch protection on `source` *instead*, and review/approve source code changes only. But this seems pretty pointless, because **if we allow Travis to push to `master`, then any approved contributor can push to `master`.** This seems like a bad idea security reasons as any maintainer with a compromised account could unilaterally override the site content. Maybe this is an acceptable tradeoff, but I think there are better options.

#### A2: Use Github Actions to push directly to `master`

For more security, it appears that we can leverage a new CI option - Github Actions. 

If [this post](https://github.community/t5/GitHub-Actions/How-to-push-to-protected-branches-in-a-GitHub-Action/m-p/36058/highlight/true#M2433) is correct, we can configure `master` to allow Github Actions to bypass the branch restrictions (by using a user key created by an admin).

As far as I can tell, this isn't insecure in the same way as using Travis because of Travis' permissions system (allowing contributors to utilise secrets throughout the repository), whereas Github Actions is more restrictive (to admins and branches).

This downside is that we'll need to write the build script for Github Actions, but I don't think that's complicated. 

Another caveat is that the build preview currently generated by Netlify might not be an accurate representation of what actually gets deployed (if there is some difference in the Github Pages build environment compare to Netlify).

### Possible Solution B: Deploy somewhere else (like Netlify)

To get around the restrictions of Github pages, we can remove the headache of having to think about pushing *any* compiled code to the Github repo. I suggest Netlify, which seems like it's designed to solve this (and other) problem(s). It will trigger a rebuild and deployment whenever a commit is added to a certain branch.

The advantage here compared to Github Pages approach is less config, less moving parts, and the build preview is what actually gets deployed (same content hash), rather than being rebuilt separately by Github Pages.

To keep the old github.io link working, we can just turn `master` into a simple redirect from etheruemclassic.github.io to etheruemclassic.org.

If we go with this I suggest we use some an admin's Netlify account.

## Conclusion

Right now it looks like, if it works the way I assume, the Github Actions approach has the same security implications as we do now (an admin can still upend everything, but maintainers cannot). It doesn't require changing the CNAME or things like that. It just needs the action to be created and an admin to generate a deploy key for the action. 

That being said I'm not familiar with Github Actions, nor do I understand it's permissions system. I'll follow up here with some testing with the Github Actions approach with a dummy repo and check that the security implications are as I assume above.

I think Netlify is also a good option here; is the least annoying thing to implement and has the least moving parts.

**Comment by hitchcott** [comment]
I've had a look into the github actions approach and unfortunately, solution A2 mentioned about isn't viable for two reasons:

1) After [testing the suggestion](https://github.community/t5/GitHub-Actions/How-to-push-to-protected-branches-in-a-GitHub-Action/m-p/30710/highlight/true#M526) for myself, the github action failed to push to the repo even with my personal access token (as the org admin / repo owner)
2) It would [open up security issues](https://github.community/t5/GitHub-Actions/How-to-push-to-protected-branches-in-a-GitHub-Action/m-p/30710/highlight/true#M526) as github actions can be created and triggered by any contributor (meaning any contributor could publish to master)

So we're back to options 1A or 2.

Either we drop security, or switch to a different provider (such as netlify, which is also free).

**Comment by bobsummerwill** [comment]
Thanks for trying!
Netlify sounds like the best option, I think.
Would you agree?

**Comment by hitchcott** [comment]
Yes, I think Netlify is the best option.

**Comment by hitchcott** [comment]
We could also apply for we could also apply for https://www.netlify.com/legal/open-source-policy/ if it's worth it, for a bit more decentralised control.

**Comment by TheEnthusiasticAs** [comment]
As long it is in the interest of the network, I am for it.

**Comment by hitchcott** [comment]
Okay, so i'm going to propose the following steps to switch to netlify:

- Set up netlify to build from `source`
- Add branch protection to `source`

At this point, I'll need help to point DNS is to netlify instead of github.

- Add a html redirect to ethereumclassic.org on `master` (so ethereumclassic.github.io redirects)
- Disable travis CI builds

**Comment by bobsummerwill** [comment]
Good plan, @hitchcott.
So PRs would be submitted against 'source' and then a preview built to some URL, and then when that has been previewed as "good" and multiple reviewers have looked at the source edits and approved, it would be merged into "source" and what happens from there?   Does that push the results back into master?

**Comment by hitchcott** [comment]
Correct @bobsummerwill ,

After the approved changes get merged into `source`, netlify will handle the build process rather than travis and deploy it to https://ethereumclassic.netlify.com/

It is currently configured to that as of now.

(by the way, I have also set it up so netlify also deploys EVERY branch in the same way such as: https://develop--ethereumclassic.netlify.com/)

I will talk to the relevant people on discord to manage the DNS switch.

**Comment by hitchcott** [comment]
Branch protection is now implemented on `source`; DNS switch is in the pipeline and now pending.

Right now netlify deploy is linked to my github account. Build logs are publicly available at: https://app.netlify.com/sites/ethereumclassic/deploys

For any reason, the system can be easily redeployed by switching DNS to a different netlify account with the following settings:

![image](https://user-images.githubusercontent.com/1696942/72775670-db6b3b00-3c41-11ea-96ea-c51fbd70a999.png)

**Comment by hitchcott** [comment]
Good news everyone!

https://ethereumclassic.org/ is now being served from netlify.

We're already effectively using this new workflow as of now. PRs to `source` should trigger a build preview and will automatically redeploy once merged. I'll test this now by merging `develop`.

I'll leave this issue open for the moment until I completely clean up this repo by removing travis and replacing `master` with a redirect.

**Comment by bobsummerwill** [comment]
🌈👌

**Comment by hitchcott** [comment]
Just gonna quickly re-open this and implement the change discussed in discord.

Instead of `source`, we'll now just have `master`; the redirect html and `CNAME` will become part of the source code.

**Comment by hitchcott** [comment]
Done.

I also had to add `.nojekyll` to remove github buld errors.

I also changed the netlify configuration to publish from `master`.
